### DISCUSSION TITLE
loadCount does not work when no relations exist

### DISCUSSION THREAD

**Discussion by ghost** [laravel/framework#52141] [Q&A] [answered]
### Laravel Version

11

### PHP Version

8.2

### Database Driver & Version

mysql 8

### Description

https://github.com/laravel/framework/blob/1225a2ff8c8eb85240ebd8b1eb5b7442eea335c5/src/Illuminate/Database/Eloquent/Collection.php#L96

Call to a member function getAttributes() on null

loadCount calls loadAggregate which is written without good practices. The null returning find is not handled properly, making the loadCount and loadExist functions not usable.

### Steps To Reproduce

Use loadCount([]) on an existing model.
or loadCount([relation]) on an existing model that does not have that relation.(update: it has the relation defined but it does not have any results in db for it)

UPDATE

![image](https://github.com/user-attachments/assets/84e29989-8e8a-4dc2-b0c4-ad6adc515cfe)
I see many un handled nulls..

**Comment by pogorelijStan** [comment]
same here

**Comment by pogorelijStan** [comment]
@marius-mcp are you sure you saved the model you calling loadCount() on? I realized I did not, and after persisting loadCount() works well, which is obvious if you think about it.

**Comment by ghost** [comment]
@pogorelijStan I found the issue when refactoring the $createdModelAndfetchedFromWritePdoDb->relationName()->count(); into createdModelAndfetchedFromWritePdoDb->loadCount(['relationName']);

**Comment by ghost** [comment]
And My specific case whas when the loadCount was called with [] array.

**Comment by macropay-solutions** [comment]
This is why we used
https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/a68df15710108ca30e8e48ebfbcf73dd307bd826/src/Http/Controllers/ResourceControllerTrait.php#L110
and not loadCount.

**Comment by ghost** [comment]
@crynobone  If this is not an issue I'm the pope...

**Reply by ghost** [reply, marked as answer]
@crynobone The issue arises only when using read and write replicas if the read replica is not synced before this query is issued agains the read replica after a create.

     $models = $this->first()->newModelQuery()
            ->whereKey($this->modelKeys())
            ->select($this->first()->getKeyName())
            ->withAggregate($relations, $column, $function)
            ->get()
            ->keyBy($this->first()->getKeyName());

Update. It might also the the case when someone uses a transaction on db that is not committed to create and save a new model and then tries to load things on it which go in read replica.


Solution from my part on a single model:
//load

        foreach ($relations as $relationName) {
            $model->{$relationName};
        }

//loadCount or loadExists

        foreach ($relations as $relationName) {
            $model->setAttribute($relationName . '_count', $model->{$relationName}()->count());
        }

**Reply by ghost** [reply]
The same is valid for load function by the way. @crynobone

**Reply by macropay-solutions** [reply]
Taking into account that read and write replicas are supported by laravel, this behavior should be stated in the documentation if this is not considered an issue @crynobone.

**Reply by ghost** [reply]
Also anything that uses $this->modelKeys() on collections will not work with composed primary keys @macropay-solutions .

**Comment by ghost** [comment]
@crynobone Bottom line is I did not investigate further when I saw the error:

    Call to a member function getAttributes() on null


I found the happy flow-ish code and wrote the issue.

This is just another example why writing code on the happy flow is not good practice.
I learned from a colleague many years ago that chaining functions calls after multi return type, without checking the return type is BAD. This is just another proof for that...

I remember I've read somewhere in here that someone wanted to return instead of null an empty model or something on collections... I guess this was the reason.
